ledger/complete/wal/checkpoint_v6_reader.go

```diff
@@ -0,0 +1,340 @@
+	}
+
+	return tries, nil
+}
+
+func filePathHeader(dir string, fileName string) string {
+	return path.Join(dir, fileName)
+}
+
+func filePathSubTries(dir string, fileName string, index int) (string, string, error) {
+	if index < 0 || index > 15 {
+		return "", "", fmt.Errorf("index must be between 1 to 16, but got %v", index)
+	}
+	subTrieFileName := fmt.Sprintf("%v.%v", fileName, index)
+	return path.Join(dir, subTrieFileName), subTrieFileName, nil
+}
+
+func filePathTopTries(dir string, fileName string) (string, string) {
+	topTriesFileName := fmt.Sprintf("%v.%v", fileName, 16)
+	return path.Join(dir, topTriesFileName), fileName
+}
+
+// TODO: to add
+func readHeader(filePath string) ([]uint32, uint32, error) {
+	return make([]uint32, 16), 0, nil
+}
+
+// func readNodeCountAndTriesCount(f *os.File) (uint64, uint16, error) {
+// 	// read the node count and tries count from the end of the file
+// 	const footerOffset = encNodeCountSize + encTrieCountSize + crc32SumSize
+// 	const footerSize = encNodeCountSize + encTrieCountSize // footer doesn't include crc32 sum
+// 	footer := make([]byte, footerSize)                     // must not be less than 1024
+// 	_, err := f.Seek(-footerOffset, io.SeekEnd)
+// 	if err != nil {
+// 		return 0, 0, fmt.Errorf("cannot seek to footer: %w", err)
+// 	}
+// 	_, err = io.ReadFull(f, footer)
+// 	if err != nil {
+// 		return 0, 0, fmt.Errorf("could not read footer: %w", err)
+// 	}
+//
+// 	return decodeFooter(footer)
+// }
+
+func readSubTriesConcurrently(dir string, fileName string, subtrieChecksums []uint32) ([][]*node.Node, error) {
+	// TODO: replace 16 with const
+	if len(subtrieChecksums) != 16 {
+		return nil, fmt.Errorf("expect subtrieChecksums to be %v, but got %v", 16, len(subtrieChecksums))
+	}
+
+	resultChs := make([]chan *resultReadSubTrie, 0, 16)
+	for i := 0; i < 16; i++ {
+		resultCh := make(chan *resultReadSubTrie)
+		go func(i int) {
+			nodes, err := readCheckpointSubTrie(dir, fileName, i, subtrieChecksums[i])
+			resultCh <- &resultReadSubTrie{
+				Nodes: nodes,
+				Err:   err,
+			}
+			close(resultCh)
+		}(i)
+		resultChs = append(resultChs, resultCh)
+	}
+
+	nodesGroups := make([][]*node.Node, 0)
+	for i, resultCh := range resultChs {
+		result := <-resultCh
+		if result.Err != nil {
+			return nil, fmt.Errorf("fail to read %v-th subtrie, trie: %w", i, result.Err)
+		}
+
+		nodesGroups = append(nodesGroups, result.Nodes)
+	}
+
+	return nodesGroups, nil
+}
+
+// subtrie file contains:
+// 1. checkpoint version // TODO
+// 2. nodes
+// 3. node count
+// 4. checksum
+func readCheckpointSubTrie(dir string, fileName string, index int, checksum uint32) ([]*node.Node, error) {
+	// TODO: read and validate checksum
+
+	filepath, _, err := filePathSubTries(dir, fileName, index)
+	if err != nil {
+		return nil, err
+	}
+	f, err := os.Open(filepath)
+	if err != nil {
+		return nil, fmt.Errorf("could not open file %v: %w", filepath, err)
+	}
+	defer func(f *os.File) {
+		f.Close()
+	}(f)
+
```

Comment:
>  since the part files don't have magic code and version, then you won't read the part file as checkpoint header file of any version by mistake.

I think different file types should contain their own unique magic id.  This helps prevent accidental collisions (a part file's data being equal to checkpoint header file's magic id by coincidence).

Also, including a version number with the magic id is useful for detecting out-of-sync versions during development/testing.

We can use a tiny sanity-check header in part files to:
- know it's a part file (magic ID for checkpoint part file)
- know version of part file format (without relying on other files in case they become out of sync)

When handling a collection of related files, we want info that can help us debug more quickly unless that is cost prohibitive.  Adding a tiny header with unique magic id + version to identify file type is cheap.

E.g. during development/testing, it would take longer to debug file checksum mismatch than a file header showing file format versions are out of sync.

Reply:
makes sense to me. [Issued up here](https://github.com/onflow/flow-go/issues/3325), will address that in separate PR